### ISSUE TITLE
Propose configurable ring buffer in merge operator

### ISSUE THREAD

**Issue by avram** [ReactiveX/RxJava#3041]
tl;dr; Propose new merge method with custom ring buffer configuration: 
```java
public final static <T> Observable<T> mergeWithBuffer(Observable<? extends Observable<? extends T>> source, long mergeBufferSize) {}
```

This is my use case, from our current production system:

Consider a firehose of events, in our case game actions. We have some low-priority events, such as game engine telemetry, and high-priority events such as purchases and installs. Our application receives the URLs of large batches of such events found on S3, and is responsible for loading them into a data warehouse. The warehouse prefers to load several hundred batches at one go, and has a fixed limit on how many loads it can perform at the same time.

The two categories of data, priority and ordinary, are to be loaded under different SLAs-- priority data should be loaded within 2 hours, and ordinary data can take up to 12 hours.

```
void process(Observable<DataSlice> inputData, Predicate<String> priority) {
	inputData
		.groupBy(i -> i.name)
		// Here our CopyRequests can accumulate inside the merge operator. We would much prefer for them to produce backpressure immediately.
		.flatMap(group -> 
			group
				// We buffer up to 500 items, waiting up to 2 hours for priority tables, or 12 hours for ordinary ones.
				.buffer(500, priority.test(group.getKey()) ? 2 : 12, TimeUnit.HOURS)
				.map(list -> new CopyRequest(group.getKey(), list))
				// We actually use a custom operator here that provides us with monitoring on the number of buffered copy requests.
				.onBackpressureBuffer()
		)
		.subscribe(new Observer<CopyRequest>() {
			//..

			@Override
			public void onNext(CopyRequest next) {
				// takes a long time writing to the warehouse.
			}

			//..
		});
}
```

Our goal is to have nearly all of the `CopyRequest` instances in per-group buffers; they instead accumulate inside the ring buffer; as each instance corresponds to large volume of data, the invisible set of requests inside the ring buffer is quite substantial.

Ultimately, we want to actually have the backpressure interact with the per-group buffering to let the groupings continue to buffer into larger groups when data warehouse writes are behind, but the first step is to reduce the in-merge buffering.

We can tune the behavior of `flatMap`'s internal `OperatorMerge`'s ring buffer with `rx.ring-buffer.size`, but that will have undesired effects elsewhere in our Rx orchestration, as it is global and must be set before any usage of the merge operator.

I think what we want is a way to do `Observable.merge()` with a specified buffer size, something like,

```
public final static <T> Observable<T> mergeWithBuffer(Observable<? extends Observable<? extends T>> source, long mergeBufferSize) {}
```

Clearly the default behavior of `merge(..)` is vital to high volume applications that can't afford the allocation of new operators with custom configurations, but a new method like `mergeWithBuffer(..)` would allow for merges where buffering is less desired.

If this is a reasonable proposal, I'm happy to start work on implementation.

**Comment by avram** [comment]
As I clarified on Twitter, the goal here is to make `Observable`s upstream of the merge more responsive to backpressure by reducing the number of items that merge will silently buffer inside its own operator.

**Comment by benjchristensen** [comment]
Buffering in operators like merge is an implementation detail to achieve non-blocking behavior. It should not be expected to pull a single value at a time like an Iterable. 

It is still not clear to me why reducing the upstream demand is wanted when the intent is to buffer or batch writes. 

What problem is the buffering inside merge causing?

**Comment by avram** [comment]
The problem is that I have operators responsible for buffering upstream of the merge, and those operators can't find out about the backpressure until 128 of their batches are already in-flight inside the merge operator.

**Comment by akarnokd** [comment]
I've been thinking about operator configuration objects so each of them can be tailored for a particular sequence:

```java
interface MergeSettings {
    int maxConcurrency();
    boolean delayErrors();
    boolean useStandardQueue();
    int innerReplenishLimit();
    int outerReplenishLimit();
    int queueCapacity();
    Queue<Object> createQueue();
    boolean useAtomicSerialization();
}
```

**Comment by davidmoten** [comment]
I've had big performance gains processing big streams with large ring buffer sizes but wanted to localize the effect so I'd be happy to see more configurability.

**Comment by benjchristensen** [comment]
> The problem is that I have operators responsible for buffering upstream of the merge, and those operators can't find out about the backpressure until 128 of their batches are already in-flight inside the merge operator.

What is the problem this is causing? Are you trying to restrict the number of batches in flight? Prevent production of data until it is needed?

It actually sounds like you want an Iterable that pulls, not an Observable that pushes.

**Comment by avram** [comment]
> What is the problem this is causing? Are you trying to restrict the number of batches in flight? Prevent production of data until it is needed?

So I have multiple upstream observables, each buffering items of a single type into larger item lists. When the downstream consumer can't keep up, I want to (1) have visibility into how many item lists of each type are in-flight and (2) be able to eventually have those buffering operators respond to backpressure by delaying emissions until the downstream is ready to consume.

I suppose I am trying to restrict the number of batches in flight, but my first goal is to be able to monitor the number of batches of each type in flight.

**Comment by avram** [comment]
And to @akarnokd's note-- such an ability to have per-merge specification of operator configuration would certainly do the trick.

**Comment by akarnokd** [comment]
2.x plans to add a bufferSize parameter to merge/flatMap that let's you control the internal buffer sizes. I don't think this will be added to 1.x due to API surface concerns.

**Comment by avram** [comment]
Sounds great! Thanks for considering the use case.
